### PR TITLE
Add setup.py

### PR DESCRIPTION
**Pull request by Pipboyguy**
Allows for a `pip install -e .[dev]`. 

This allows for:
- better linting
- beginner friendly IDE assistance

### CODE REVIEW

**Review by Weves — COMMENTED** [review]
Thanks for the change! The initial setup can definitely be made easier and we really appreciate you putting in the time and effort to help out here. 

As far as I know, `setup.py` is mostly considered legacy (e.g. if you look at the most up to date `pip` documentation, https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#, then you'll see `setup.py` marked as legacy). Is there a specific reason for using this over something more modern like `pyproject.toml` + something like the following for the `setup.py`:

```
#!/usr/bin/env python

import setuptools

if __name__ == "__main__":
    setuptools.setup()
```

Or even more alternatively, using poetry (https://python-poetry.org/). I've never personally used poetry, but I've heard good things about it.

**Comment by Pipboyguy** [comment]
@Weves updated to `pyproject.toml`

**Comment by yuhongsun96** [comment]
Hey @Pipboyguy,
Couple things if you don't mind:
1. [important] We don't want to duplicate the requirements, if we are moving towards a package management util, we should remove the files under requirements. This also means replacing the Docker builds which pull in the reqs and installs them into the container.
2. [optional] I think Poetry is pretty good as well, maybe you can do something like:
```
[tool.poetry]
name = "Danswer"
version = "0.1.0"
description = "Opensource Enterprise Question Answering"
license = "MIT"
# whatever else

[tool.poetry.dependencies]
python = "^3.8"
alembic = "1.10.4"
asyncpg = "0.27.0"
atlassian-python-api = "3.37.0"
beautifulsoup4 = "4.12.0"
# other dependencies

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
# whatever else
```
3. [minor] I know we don't do a great job of this currently, but it's also better to specify a min version of the dependencies rather than to pin it to an exact one.

If all this is too much hassle, don't worry about it, there are higher value items if you're looking to contribute, feel free to DM us.
But anyway, at the very least, we don't want to duplicate reqs in multiple places.

File: backend/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+[tool.setuptools]
+packages = ["danswer"]
+
+[project]
+name = "danswer"
+requires-python = ">=3.7"
+license = {text = "MIT License"}
+version = "0.1"
+dependencies = [
```
Comment (yuhongsun96) [inline backend/pyproject.toml:13]: Would want to avoid duplicating dependencies

File: backend/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+       "google-auth-oauthlib==1.0.0",
+       "httpcore==0.16.3",
+       "httpx==0.23.3",
+       "httpx-oauth==0.11.2",
+       "jira==3.5.1",
+       "Mako==1.2.4",
+       "nltk==3.8.1",
+       "openai==0.27.6",
+       "playwright==1.32.1",
+       "psycopg2==2.9.6",       "black==23.3.0",
```
Comment (yuhongsun96) [inline backend/pyproject.toml:33]: minor neatness fix

File: backend/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+       "types-html5lib==1.1.11.13",
+       "types-psycopg2==2.9.21.10",
+       "types-regex==2023.3.23.1",
+       "types-requests==2.28.11.17",
+       "types-urllib3==1.26.25.11",
+]
+cdk = [
+    "aws-cdk-lib>=2.0.0",
+    "constructs>=10.0.0"
+]
```
Comment (yuhongsun96) [inline backend/pyproject.toml:73]: [nit] newline

File: backend/setup.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env python
+
+import setuptools
+
+if __name__ == "__main__":
+    setuptools.setup()
```
Comment (yuhongsun96) [inline backend/setup.py:6]: [nit] newline

**Review by yuhongsun96 — CHANGES_REQUESTED** [review]
TLDR, let's avoid duplicate listing of requirements. Ideally move to poetry and make necessary changes to build processes.
But if it's too much of a hassle, it's ok to close this, there are higher priority stuff we can suggest!

Thanks!
